syntax = "proto3";

package DiSy;

message DirectoryMetadata {
    string relative_path = 1;
}

message FileMetadata {
    string relative_path = 1;
    int64 last_modified_time = 2;
    string hash = 3;
}

message DirTree {
    map<string, DirectoryMetadata> directories = 1;
	map<string, FileMetadata> files = 2;
}

message UpdateRequest {
    int64 client_id = 1;
    int64 time = 2;
    DirTree dir_tree = 3;
}

message UpdateResponse {
    DirTree uploads = 1;
    DirTree downloads = 2;
}

message Empty{
}

message GetNewIdResponse {
    int64 id = 1;
}

service Server {
    rpc GetNewId(Empty) returns (GetNewIdResponse);
    rpc Update(UpdateRequest) returns (UpdateResponse);
}